src/service/video/docking.js

```diff
@@ -0,0 +1,1418 @@
+
+    this.placeAt_(video, x, y, scale, step);
+
+    this.setCurrentlyDocked_(video, posX, posY, step);
+
+    this.getDockingTimeout_().trigger(DOCKING_TIMEOUT, video);
+  }
+
+  /**
+   * @param  {!VideoOrBaseElementDef} video
+   * @return {boolean}
+   * @private
+   */
+  ignoreDueToDismissal_(video) {
+    if (this.lastDismissed_ != video || this.lastDismissedPosY_ !== null &&
+        !this.positionMatchesScroll_(this.lastDismissedPosY_)) {
+      return false;
+    }
+    if (!this.isVisible_(video.element, FLOAT_TOLERANCE)) {
+      this.resetDismissed_();
+    }
+    return true;
+  }
+
+  /** @private */
+  resetDismissed_() {
+    this.lastDismissed_ = null;
+    this.lastDismissedPosY_ = null;
+  }
+
+  /**
+   * Prevents jump when the transition was timed out before user finished
+   * scrolling component out of view.
+   * @return {boolean}
+   */
+  ignoreDueToTransitionEnd_(step) {
+    return this.hasTransitionCompleted_(step) &&
+      this.currentPositionMatchesScroll_();
+  }
+
+  /**
+   * @param {!AmpElement} element
+   * @return {number}
+   * @private
+   */
+  calculateStep_(element) {
+    const {intersectionRatio} = element.getIntersectionChangeEntry();
+    return 1 - Math.pow(intersectionRatio, 3);
+  }
+
+  /**
+   * @param {number} step
+   * @return {number}
+   */
+  calculateTransitionDuration_(step) {
+    const maxAutoTransitionDurationMs = 500;
+    if (!this.currentlyDocked_) {
+      // Don't animate first frame. Browsers sometimes behave weirdly and use
+      // a stale transform value, thus causing it to visually jump.
+      return 0;
+    }
+    const remaining = Math.abs(step - this.currentlyDocked_.step);
+    return remaining * maxAutoTransitionDurationMs;
+  }
+
+  /**
+   * @param {number} x
+   * @param {number} y
+   * @param {number} scale
+   * @return {boolean}
+   */
+  alreadyPlacedAt_(x, y, scale) {
+    return !!this.placedAt_ &&
+        this.placedAt_.x == x &&
+        this.placedAt_.y == y &&
+        this.placedAt_.scale == scale;
+  }
+
+  /**
+   * @param {!VideoOrBaseElementDef} video
+   * @param {number} x
+   * @param {number} y
+   * @param {number} scale
+   * @param {number} step in [0..1]
+   * @private
+   */
+  placeAt_(video, x, y, scale, step, optTransitionDurationMs = null) {
+    if (this.alreadyPlacedAt_(x, y, scale)) {
+      return;
+    }
+
+    const transitionDurationMs = optTransitionDurationMs ?
+      dev().assertNumber(optTransitionDurationMs) :
+      this.calculateTransitionDuration_(step);
+
+    const {width, height} = video.getLayoutBox();
+
+    this.placedAt_ = {x, y, scale};
+
+    const transitionTiming =
+        // Auto-transitions are supposed to smooth-out PositionObserver
+        // frequency, so it makes sense to use 'linear'. When the transition
+        // duration is otherwise larger, 'ease-in' looks much nicer.
+        transitionDurationMs > 200 ? 'ease-in' : 'linear';
+
+    const positioningStyles = {
+      'width': px(width),
```

Comment:
do we still need to set width/height given scale is being set now? would be nice to avoid if at all possible since it is causes relayout (although just relayout of video box) if not avoidable, let's add "will-change: width, height, transform" to the CSS of the docking container.

Reply:
Unfortunately yes, because we need to set the aspect ratio for each sized element.

Made a small optimization to set/reset dimensions only for first and last frames and added `will-change`.